metadata-ingestion/src/datahub/ingestion/source/gc/datahub_gc.py

```diff
@@ -140,36 +161,61 @@ def create(cls, config_dict, ctx):
     def get_workunit_processors(self) -> List[Optional[MetadataWorkUnitProcessor]]:
         return [partial(auto_workunit_reporter, self.get_report())]
 
+    @contextmanager
+    def _timed_execution(self, operation_name: str) -> Generator[None, None, None]:
+        """Context manager to handle timing and error reporting of cleanup operations."""
+        start_time = time.time()
+        try:
+            yield
+            elapsed = time.time() - start_time
+            self.report.record_success(operation_name, elapsed)
+        except Exception as e:
+            elapsed = time.time() - start_time
+            self.report.record_failure(operation_name, str(e), elapsed)
+
+    def _execute_cleanup(self, task: CleanupTask) -> Generator[Any, None, None]:
+        """Execute a cleanup operation if the condition is met."""
+        if not task.condition:
+            return
+
+        with self._timed_execution(task.name):
+            result = task.lambda_func()
+            if task.is_generator:
+                yield from result
+
     def get_workunits_internal(
         self,
     ) -> Iterable[MetadataWorkUnit]:
-        if self.config.cleanup_expired_tokens:
-            try:
-                self.revoke_expired_tokens()
-            except Exception as e:
-                self.report.failure("While trying to cleanup expired token ", exc=e)
-        if self.config.truncate_indices:
-            try:
-                self.truncate_indices()
-            except Exception as e:
-                self.report.failure("While trying to truncate indices ", exc=e)
-        if self.soft_deleted_entities_cleanup:
-            try:
-                self.soft_deleted_entities_cleanup.cleanup_soft_deleted_entities()
-            except Exception as e:
-                self.report.failure(
-                    "While trying to cleanup soft deleted entities ", exc=e
-                )
-        if self.execution_request_cleanup:
-            try:
-                self.execution_request_cleanup.run()
-            except Exception as e:
-                self.report.failure("While trying to cleanup execution request ", exc=e)
-        if self.dataprocess_cleanup:
-            try:
-                yield from self.dataprocess_cleanup.get_workunits_internal()
-            except Exception as e:
-                self.report.failure("While trying to cleanup data process ", exc=e)
+        cleanup_tasks: List[CleanupTask] = [
```

Comment:
Is the CleanupTask here for only timing?

Reply:
I think it could be more helpful if this list could be specified in the config and not here and therefore we could initialize all the CleanupTask purely from the config.